File: cs/ch16/Money/Money.cs
using System.Net;

namespace TheSoftwareGorilla.TDD.Money;

public class Money : Expression
{

    public static Money Dollar(int amount)
    {
        return new Money(amount, "USD");
    }

    public static Money Franc(int amount)
    {
        return new Money(amount, "CHF");
    }

    // Shout out to my South African friends!
    public static Money Rand(int amount)
    {
        return new Money(amount, "ZAR");
    }

    public int Amount { get; }

    public string Currency { get; }

    public Money(int amount, string currency)
    {
        Amount = amount;
        Currency = currency;
    }

    public Expression Times(int multiplier)
    {
        return new Money(Amount * multiplier, Currency);
    }

    public Expression Plus(Expression addend)
    {
        return new Sum(this, addend);
    }

    public Money Reduce(Bank bank, string to)
    {
        int rate = bank.Rate(Currency, to);
        return new Money(Amount / rate, to);
    }

    public override bool Equals(object? obj)
    {
        return obj is Money money 
            && Amount == money.Amount 
            && Currency == money.Currency;
    }

    public override string ToString()
    {
        return Amount + " " + Currency;
    }


}

File: java/ch17-02-plus/src/main/java/com/thesoftwaregorilla/tdd/money/Money.java
package com.thesoftwaregorilla.tdd.money;


public class Money implements Expression {

    public static Money dollar(int amount) {
        return new Money(amount, "USD");
    }

    public static Money franc(int amount) {
        return new Money(amount, "CHF");
    }

    // Shout out to all my South African friends!
    public static Money rand(int amount) {
        return new Money(amount, "ZAR");
    }


    private final int amount;
    private final String currency;

    Money(int amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public Expression times(int multiplier) {
        return new Money(getAmount() * multiplier, getCurrency());
    };

    public Expression plus(Expression addend) {
        return new Sum(this, addend);
    }

    public Money reduce(Bank bank, String to) {
        int rate = bank.rate(getCurrency(), to);
        if (rate == 0) {
            throw new ArithmeticException("Exchange rate not available");
        }
        return new Money(getAmount() / rate, to);
    }

    @Override
    public boolean equals(Object object) {
        Money money = (Money) object;
        return getAmount() == money.getAmount() && getCurrency().equals(money.getCurrency());
    }

    @Override
    public String toString() {
        return amount + " " + currency;
    }

    // See my note in the MoneyTest class. I added a getter for amount because I had a constructor test.
    public int getAmount() {
        return amount;
    }

    // I changed the name of this method from "currency" to "getCurrency" to keep consistent with Java standards.
    // I'm not sure these standards existed when Kent wrote the book.
    public String getCurrency() {
        return currency;
    }

}
